pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,356 @@
+	workingDirs = []string{
+		api.UploadScripts,
+		api.Source,
+		api.DefaultScripts,
+		api.UserScripts,
+	}
+)
+
+// Dockerfile builders produce a Dockerfile rather than an image.
+// Building the dockerfile w/ the right context will result in
+// an application image being produced.
+type Dockerfile struct {
+	fs               fs.FileSystem
+	uploadScriptsDir string
+	uploadSrcDir     string
+	sourceInfo       *git.SourceInfo
+	result           *api.Result
+	ignorer          build.Ignorer
+}
+
+// New creates a Dockerfile builder.
+func New(config *api.Config, fs fs.FileSystem) (*Dockerfile, error) {
+	return &Dockerfile{
+		fs: fs,
+		// where we will get the assemble/run scripts from on the host machine,
+		// if any are provided.
+		uploadScriptsDir: api.UploadScripts,
+		uploadSrcDir:     api.Source,
+		result:           &api.Result{},
+		ignorer:          &ignore.DockerIgnorer{},
+	}, nil
+}
+
+// Build produces a Dockerfile that when run with the correct filesystem
+// context, will produce the application image.
+func (builder *Dockerfile) Build(config *api.Config) (*api.Result, error) {
+
+	// Handle defaulting of the configuration that is unique to the dockerfile strategy
+	if strings.HasSuffix(config.AsDockerfile, string(os.PathSeparator)) {
+		config.AsDockerfile = config.AsDockerfile + string(os.PathSeparator) + "Dockerfile"
+	}
+	if len(config.AssembleUser) == 0 {
+		config.AssembleUser = "1001"
+	}
+	if !user.IsUserAllowed(config.AssembleUser, &config.AllowedUIDs) {
+		return nil, s2ierr.NewUserNotAllowedError(config.AssembleUser, false)
+	}
+
+	dir, _ := filepath.Split(config.AsDockerfile)
+	config.PreserveWorkingDir = true
+	config.WorkingDir = dir
+
+	if config.BuilderImage == "" {
+		builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return builder.result, errors.New("builder image name cannot be empty")
+	}
+
+	if err := builder.Prepare(config); err != nil {
+		return builder.result, err
+	}
+
+	if err := builder.CreateDockerfile(config); err != nil {
+		builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
+			utilstatus.ReasonDockerfileCreateFailed,
+			utilstatus.ReasonMessageDockerfileCreateFailed,
+		)
+		return builder.result, err
+	}
+
+	builder.result.Success = true
+
+	return builder.result, nil
+}
+
+// CreateDockerfile takes the various inputs and creates the Dockerfile used by
+// the docker cmd to create the image produced by s2i.
+func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
+	glog.V(4).Infof("Constructing image build context directory at %s", config.WorkingDir)
+	buffer := bytes.Buffer{}
+
+	if len(config.ImageWorkDir) == 0 {
+		config.ImageWorkDir = "/opt/app-root/src"
+	}
+
+	imageUser := config.AssembleUser
+
+	// where files will land inside the new image.
+	// TODO: get this from the image metadata
```

Comment:
is this //TODO still relevant given what you have discovered?  If so, was there a reason you didn't do it in this PR?